Add Course List block to My Courses on page creation

Imran92 opened the pull request:
Implements https://github.com/Automattic/sensei/issues/6341

### Changes proposed in this Pull Request

- Make sure you have an existing installation of Sensei having this branch checked out; keep it as it is
- Create a fully new local WP site containing this branch
- Install the Sensei Plugin in the new installation and complete the onboarding
- Create multiple Courses and publish them
- Create a student user
- Login as the user, go to my courses page
- Make sure you see the course list block instead of the Student Courses block
- Make sure the filter and paginations are working
- Now go to the wp installation with Sensei that you already had
- Go to the My Courses page
- Make sure it appears as before, with Student Courses block

Imran92 committed:
Add course list block in newly added my courses page

Imran92 committed:
Add test for my courses page having course list block

m1r0 commented:
Hey Imran, happy new year! 🎆 🥂 🤝 🙂 

I'm not sure if this is related to your PR, but I've noticed some issues with the Storefront theme. I haven't tested on many other themes, but it might be a good idea.

![image](https://user-images.githubusercontent.com/1612178/210267123-3df782ed-cbdc-4006-a1b6-aa0161321600.png)

Another thing I've noticed is that the login form is no longer accessible. If you are not logged in and go to My Courses, you see a list of all courses. If you try to enroll, you get redirected back to My Courses but you don't see the login form. Do you think it makes sense to show the login form on the My Courses page as before?

Imran92 committed:
Add login form setting in Course List block side panel

Imran92 committed:
Conditional render login form instead Course List block content

Imran92 committed:
Load the Course List block functions

Imran92 committed:
Update inner blocks with login form and list view

Imran92 commented:
Thanks a lot for checking @m1r0 ❤️ I've updated several parts of the PR, can you kindly give it a check again? Thanks!

donnapep commented:
I left some comments on the [P2 post](https://nexusp2.wordpress.com/2023/01/04/course-list-block-in-my-courses-and-archive/) re: the login form.

Imran92 committed:
Render login form in my courses page for course list block

Imran92 committed:
Add default options in course filter to preselect specific option

Imran92 committed:
Merge branch 'trunk' into add/course-list-block-in-my-courses-page

Imran92 commented:
> I left some comments on the [P2 post](https://nexusp2.wordpress.com/2023/01/04/course-list-block-in-my-courses-and-archive/) re: the login form.

Thanks a lot for the comments! Updated as per our discussions there [here](https://github.com/Automattic/sensei/pull/6343/commits/a7ef83b1c601baa0cc26f7a2d58639340b2976ea)

m1r0 commented on includes/blocks/course-list/class-sensei-course-list-block.php:
```diff
@@ -0,0 +1,58 @@
+			array_key_exists( 'className', $block['attrs'] ) &&
+			false !== strpos( $block['attrs']['className'], 'wp-block-sensei-lms-course-list' );
+
+		$is_my_courses_page = isset( $post ) && is_page() && intval( Sensei()->settings->get( 'my_course_page' ) ) === $post->ID;
```
Do you think the following check might be enough and a bit easier to read? You could also clean up the global in this case.

```suggestion
		$is_my_courses_page = get_the_ID() === (int) Sensei()->settings->get( 'my_course_page' );
```

Imran92 replied:
Thanks for the suggestion, updated it here https://github.com/Automattic/sensei/pull/6343/commits/b17e2c624a071fb7b721d18cc21ead880430d44f

m1r0 replied:
Nice! One small thing - you can also clean up the `global $post;` located a few lines up.

Imran92 replied:
Oh, just noticed that it got left out from my commit, removed [here](https://github.com/Automattic/sensei/pull/6343/commits/294148e85e03a340f3f7507e0866c16a25caedfe)

m1r0 reviewed (commented):
Hey, I've tested the login form and it works as described. Nice! 🎉  Left one really minor suggestion.

But my main concern here is that I feel this block still doesn't look that good on most themes:

Astra
![image](https://user-images.githubusercontent.com/1612178/211640818-4ea321bb-fb8b-4d1a-ab8c-4f8241a05935.png)

Divi
![image](https://user-images.githubusercontent.com/1612178/211641030-9cd46a09-de52-4b45-8026-1d98697f06b6.png)

Hello Elementor
![image](https://user-images.githubusercontent.com/1612178/211641942-ae4377b5-25b4-45c8-8064-e099a618c4da.png)

Maybe we should polish it a bit before making it the default?

cc @donnapep

m1r0 commented:
Are the E2E tests failing related to this PR?

donnapep commented:
> Maybe we should polish it a bit before making it the default?

Yes, I think so. I wouldn't worry about Hello Elementor, but it should definitely look good on Course, Divi and Astra.

Imran92 committed:
Fix designs on Astra, Divi and Course theme

Imran92 committed:
Fix logic check to load login page

Imran92 commented:
Thanks for taking a look @m1r0 and @donnapep <3 I've updated them a bit, and now hopefully it'll look decent across themes. Here are some screenshots from different themes

**Course**
![Screen Shot 2023-01-12 at 3 27 49 AM](https://user-images.githubusercontent.com/6820724/211923102-13157921-a910-4b8b-b5f1-332d78db7294.png)

**Astra**
![Screen Shot 2023-01-12 at 3 47 29 AM](https://user-images.githubusercontent.com/6820724/211924895-9d5509b0-607c-4735-b241-a8212fa912db.png)

**Divi**
![Screen Shot 2023-01-12 at 3 30 21 AM](https://user-images.githubusercontent.com/6820724/211924938-e7b42c9a-7326-4501-8bd9-9427b6c07a59.png)

Imran92 committed:
Align course categories left

donnapep commented:
I did some testing on the Course theme and here's what I found:

- My Courses displays pagination and shows 3 courses at a time, even though I've set Settings > Reading > Blog pages show at most = 10. Even if I delete the page, change that value to 2, and run the wizard again, it still shows 3 posts. This doesn't happen on the archive page, which works great with that setting.
- When filtering My Courses page and I'm on page 2, I don't see any results. Instead, I see a blank page. Maybe we should redirect back to page 1 after filtering? For course archive, I get a 404. (Feel free to log this one as a separate issue and address it later if this is an existing issue with the Course List Filter block.)

Imran92 committed:
Merge branch 'trunk' into add/course-list-block-in-my-courses-page

Imran92 commented:
Thanks for taking a look @donnapep !

> My Courses displays pagination and shows 3 courses at a time, even though I've set Settings > Reading > Blog pages show at most = 10. Even if I delete the page, change that value to 2, and run the wizard again, it still shows 3 posts. This doesn't happen on the archive page, which works great with that setting.

My Courses page probably doesn't consider this setting, I tried with Student Courses block as it had before and it didn't consider the number of posts set in the reader. That setting works with Archive pages global query is used there. I've just checked that the `Query Loop` block does not consider this setting either. So we'll have to find a way to make it work if we want Query Loop to support it. As this is a new behavior, do you think we should create a separate ticket for this and try to implement it there? And if we do, should that be one with priority for our release?

> When filtering My Courses page and I'm on page 2, I don't see any results. Instead, I see a blank page. Maybe we should redirect back to page 1 after filtering? For course archive, I get a 404. (Feel free to log this one as a separate issue and address it later if this is an existing issue with the Course List Filter block.)

Very good idea, sounds like a nice improvement for the Course List Filter block. I'll create a ticket for it and implement it under that. Thanks you!

Imran92 committed:
Merge branch 'trunk' into add/course-list-block-in-my-courses-page

Imran92 committed:
Remove unused global post declaration

donnapep commented:
> As this is a new behavior, do you think we should create a separate ticket for this and try to implement it there? And if we do, should that be one with priority for our release?

Nope, you're right. That setting only applies to archive pages, so we don't need to make it work for My Courses. A related problem I'm seeing though is that we're setting `per_page` to 3 [here](https://github.com/Automattic/sensei/pull/6343/files#diff-cbbc67b96a2d760b36f6b61bb11ec126f11a69189cf186a8e0b9452615fb665dR163). It seems to me as though it would be better to show all courses on the same page. I think grid view may be setting a limit as well. Perhaps we should show however many courses the legacy My Courses page shows?

A couple new issues I found:
- Although the _Active_ filter is set by default, when I view My Courses it's showing all courses, not just the active ones.
- If I have a paid course on the second page and I click the _Buy Course_ button, I don't see the notification that the course was added to the cart.

Imran92 committed:
Make param key a const

Imran92 committed:
Add function to generate param for default filter

Imran92 committed:
Add function to get default options, generate params

Imran92 committed:
Add tests for default filtering of course list block

Imran92 committed:
Update per page setting to ten by default

Imran92 committed:
Merge branch 'trunk' into add/course-list-block-in-my-courses-page

Imran92 committed:
Take to first page after filtering

Imran92 commented:
Thanks a lot for reviewing @donnapep ! 

> Although the Active filter is set by default, when I view My Courses it's showing all courses, not just the active ones.

I've updated the code [here](https://github.com/Automattic/sensei/pull/6343/commits/7f6100f39c04e57629811e3e246c3e67c16643c3) and [here](https://github.com/Automattic/sensei/pull/6343/commits/c59c2e5f0c9b1ffe771278c37139122955a5cd20), now the frontend will filter considering the default value.

> If I have a paid course on the second page and I click the Buy Course button, I don't see the notification that the course was added to the cart.

This issue was due to the Course Purchase button removing all query params from URL, it was the user to the first page every time and as the added course wasn't there, the notice wasn't shown. I've fixed it [here](https://github.com/Automattic/sensei-pro/pull/2052) in sensei-pro.

> we're setting per_page to 3 [here](https://github.com/Automattic/sensei/pull/6343/files#diff-cbbc67b96a2d760b36f6b61bb11ec126f11a69189cf186a8e0b9452615fb665dR163). It seems to me as though it would be better to show all courses on the same page. I think grid view may be setting a limit as well. Perhaps we should show however many courses the legacy My Courses page shows?

I've updated the page pattern [here](https://github.com/Automattic/sensei/pull/6343/commits/7e08a1c9c57f73ada5be5ce7ca5b6887268e6cc8) to show the same number of courses as the legacy block.

As for post per page, Query Loop block by default does not allow showing all courses, even if we provide no value for `perPage`, it'll still add a limit. One way around this can be setting the `perPage` value very high, like 1000, so that it becomes unlikely to go past that. Should it be alright if we do that?

The limit with Grid comes from the default grid pattern we created. We can increase it if we want. The catch will be users adding the pattern anywhere will now have that increased number. We can do it too if that's okay.

> When filtering My Courses page and I'm on page 2, I don't see any results. Instead, I see a blank page. Maybe we should redirect back to page 1 after filtering

Updated it [here](https://github.com/Automattic/sensei/pull/6343/commits/482b62f61eec281adc7f2e49c01c1df8507e0494), now when filtering, we'll be taken to the first page. I'll make it work for Archive too in the Archive PR.

Imran92 committed:
Refactor filter block content method

donnapep commented:
> As for post per page, Query Loop block by default does not allow showing all courses, even if we provide no value for perPage, it'll still add a limit. One way around this can be setting the perPage value very high, like 1000, so that it becomes unlikely to go past that. Should it be alright if we do that?

Showing 10 for the list view and 12 for the grid view as it works now looks good, so we can leave it as is. 👍🏻

donnapep commented on tests/unit-tests/blocks/test-class-sensei-course-list-filter-block.php:
```diff
@@ -322,4 +322,41 @@ public function testCourseFilterBlock_WhenFilteredForCompletedFeaturedAndCategor
+		$this->assertStringNotContainsString( $this->course2->post_title, $result );
+	}
+
+	public function testCourseFilterBlock_WhenFilteredForActiveByDefault_ShowsOnlyTheFeaturedCourses() {
```
```suggestion
	public function testCourseFilterBlock_WhenFilteredForActiveByDefault_ShowsOnlyTheActiveCourses() {
```

donnapep commented on includes/admin/class-sensei-setup-wizard-pages.php:
```diff
@@ -98,9 +98,84 @@ private function get_learner_courses_page_template() {
+							'displayLayout' => [
+								'type' => 'list',
+							],
+							'className'     => 'wp-block-sensei-lms-course-list wp-block-sensei-lms-course-list\u002d\u002dis-list-view',
```
Do we need the Unicode characters here?

Imran92 replied:
Thanks, removed it [here](https://github.com/Automattic/sensei/pull/6343/commits/1ca69e3f383f8b68a25dd34bb5d762e9bb31f3fa)

donnapep commented on includes/blocks/course-list/class-sensei-course-list-block.php:
```diff
@@ -0,0 +1,56 @@
+	 *
+	 * @return string
+	 */
+	public function render_login_form_if_applicable( $block_content, $block ) {
```
We should mark this as `@private`. You also usually see functions like this using a naming convention like `maybe_render_login_form`.

Imran92 replied:
Absolutely, updated the name and annotation [here](https://github.com/Automattic/sensei/pull/6343/commits/bd70cd1c70701b7e7fe5c632d80e2de2a7b738a2)

donnapep commented on includes/blocks/course-list/class-sensei-course-list-block.php:
```diff
@@ -0,0 +1,56 @@
+		}
+
+		$is_course_list_block = array_key_exists( 'query', $block['attrs'] ) &&
+			'course' === $block['attrs']['query']['postType'] &&
```
Should we account for the possibility of `postType` also not existing?

Imran92 replied:
Very good observation, added some checks for it [here](https://github.com/Automattic/sensei/pull/6343/commits/590efcbdac8d4f11b5685b5233c251fbc69b2088)

donnapep commented on includes/blocks/course-list/class-sensei-course-list-block.php:
```diff
@@ -0,0 +1,56 @@
+	}
+
+	/**
+	 * Replaces content of Course List block when Show Login Form is set and user is logged out.
```
I think we may have removed the "Show Login Form" setting?

Imran92 replied:
Thanks! Updated the comment [here](https://github.com/Automattic/sensei/pull/6343/commits/80deb082e80904abee15dfcee5af7d2b6c40002b)

donnapep reviewed (dismissed):
Looks good! Left a few minor comments.

Imran92 committed:
Update test function name for default active courses

Co-authored-by: Donna Peplinskie <donnapep@gmail.com>

Imran92 committed:
Remove unicode characters from classname

Imran92 committed:
Update function name and add private annotation

Imran92 committed:
Add checks for post type not existing

Imran92 committed:
Update doc comment for login form render method

Imran92 committed:
Deprecate learner courses hook and add my courses hook

Imran92 commented:
Updated the hook name as per the conversation [here](https://github.com/Automattic/sensei/pull/6354#discussion_r1074018636)

donnapep commented:
I get the same fatal error as on the course archive page when I don't select a pattern, but I assume [this fix](https://github.com/Automattic/sensei/commit/76b37d4bab1719388cef8b719e536fca1631d7e2) will resolve it?

donnapep commented on includes/admin/class-sensei-setup-wizard-pages.php:
```diff
@@ -155,33 +141,62 @@ private function get_learner_courses_page_template() {
+		 *
+		 * @return {array} Blocks array.
+		 */
+		$blocks = apply_filters( 'sensei_learner_courses_page_template', $blocks );
```
Oh, whoops! I missed that this was an existing hook! To deprecate a hook, we need to use `apply_filters_deprecated`, but I think it would be easier to just revert this commit and avoid another review. 🙂

Imran92 replied:
Oh! Got it, thanks! Reverted it here https://github.com/Automattic/sensei/pull/6343/commits/3ddc975a3fad775e68c1cd1a340d8c326bf6af20

Imran92 committed:
Revert "Deprecate learner courses hook and add my courses hook"

This reverts commit ac791b7c711a453bbd9a4b911522c982ca83b14e.

Imran92 commented:
> I get the same fatal error as on the course archive page when I don't select a pattern, but I assume [this fix](https://github.com/Automattic/sensei/commit/76b37d4bab1719388cef8b719e536fca1631d7e2) will resolve it?

Yeah it'll fix this one too

donnapep reviewed (approved):
🥳
